src/ADAL.PCL.Desktop/Native/NCryptNative.cs

```diff
@@ -0,0 +1,877 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.ConstrainedExecution;
+using System.Runtime.InteropServices;
+using System.Security;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Win32.SafeHandles;
+
+namespace Microsoft.IdentityModel.Clients.ActiveDirectory.Native
+{
+    //
+    // Public facing interop enumerations
+    //
+
+    /// <summary>
+    ///     Algorithm classes exposed by NCrypt
+    /// </summary>
+    [Flags]
+    public enum NCryptAlgorithmOperations
+    {
+        None = 0x00000000,
+        Cipher = 0x00000001,                        // NCRYPT_CIPHER_OPERATION
+        Hash = 0x00000002,                          // NCRYPT_HASH_OPERATION
+        AsymmetricEncryption = 0x00000004,          // NCRYPT_ASYMMETRIC_ENCRYPTION_OPERATION
+        SecretAgreement = 0x00000008,               // NCRYPT_SECRET_AGREEMENT_OPERATION
+        Signature = 0x00000010,                     // NCRYPT_SIGNATURE_OPERATION
+        RandomNumberGeneration = 0x00000020,        // NCRYPT_RNG_OPERATION
+    }
+
+    /// <summary>
+    ///     Native wrappers for ncrypt CNG APIs.
+    ///     
+    ///     The general pattern for this interop layer is that the NCryptNative type exports a wrapper method
+    ///     for consumers of the interop methods.  This wrapper method puts a managed face on the raw
+    ///     P/Invokes, by translating from native structures to managed types and converting from error
+    ///     codes to exceptions.
+    /// </summary>
+    internal static class NCryptNative
+    {
+        //
+        // Enumerations
+        //
+
+        /// <summary>
+        ///     Well known key property names
+        /// </summary>
+        internal static class KeyPropertyName
+        {
+            internal const string Length = "Length";                // NCRYPT_LENGTH_PROPERTY
+        }
+
+        /// <summary>
+        ///     NCrypt algorithm classes
+        /// </summary>
+        internal enum NCryptAlgorithmClass
+        {
+            None = 0x00000000,
+            AsymmetricEncryption = 0x00000003,                      // NCRYPT_ASYMMETRIC_ENCRYPTION_INTERFACE
+            SecretAgreement = 0x00000004,                           // NCRYPT_SECRET_AGREEMENT_INTERFACE
+            Signature = 0x00000005,                                 // NCRYPT_SIGNATURE_INTERFACE
+        }
+
+        /// <summary>
+        ///     Enum for some SECURITY_STATUS return codes
+        /// </summary>
+        internal enum ErrorCode
+        {
+            Success = 0x00000000,                                   // ERROR_SUCCESS
```

Comment:
Error code = 0 implies success. It is the standard return code strategy. This code was borrowed  from CLR security project.